Added Dockerfile for cross compilation

Binozo opened the pull request:
Hey there 👋

I added a Dockerfile which allows to cross compile from Windows x64 or Linux x64 to both Windows x64 and Arm64.

Also added some instructions to the README

sappho192 commented:
Hi @Binozo ! How have you been? 🤗

Thank you for the contribution, I can't believe you've find a way to build ARM64 artifact using Docker! Hats off to you 🤠

Although the code seems perfect, would you mind if I have a time to test in my environment?
The test would be done in this weekend.

Binozo commented:
Hi @sappho192 all good here, thanks 😊
How are you doing?

Yes sure take your time to test it, finally got it to work 🙂

After I submit this PR I saw issue #17 and decided to include a Linux x64/arm64 build in the Dockerfile because I am already at it 🧑‍💻

Cross compilation for linux already works, tomorrow I will make some last tweaks to include the complete nuget packaging.

Warm greetings from sunny 🇩🇪🍻

Binozo commented:
Soo finally I got everything right 😄

Reversed the powershell script modifications because they are no longer needed 🙂

Have a nice weekend

Binozo commented:
Also there is no need to run the powershell scripts when you build the project with Docker. Just adjust the version in `NATIVE_LIB_VERSION.txt` and it will work out of the box 🧑‍💻

sappho192 commented:
Hi @Binozo , 
I've tried to run the Dockerfile to build the artifacts, but it was failed with following logs:

```
PS D:\REPO\Tokenizers.DotNet> docker build --tag 'tokenizers.dotnet' .
[+] Building 167.5s (41/48)                                                                   docker:desktop-linux 
 => [internal] load build definition from Dockerfile                                                          0.0s 
 => => transferring dockerfile: 3.04kB                                                                        0.0s 
 => [internal] load metadata for docker.io/mstorsjo/llvm-mingw:20250319                                       2.8s 
 => [internal] load metadata for docker.io/library/rust:1.85.0-bookworm                                       2.5s 
 => [auth] library/rust:pull token for registry-1.docker.io                                                   0.0s 
 => [auth] mstorsjo/llvm-mingw:pull token for registry-1.docker.io                                            0.0s 
 => [internal] load .dockerignore                                                                             0.0s 
 => => transferring context: 57B                                                                              0.0s 
 => [winarm64  1/10] FROM docker.io/mstorsjo/llvm-mingw:20250319@sha256:ab866cbff8b4f2131ee423742e33a1804a7  25.0s 
 => => resolve docker.io/mstorsjo/llvm-mingw:20250319@sha256:ab866cbff8b4f2131ee423742e33a1804a7a7c3e5c93b9e  0.0s 
 => => sha256:9cb31e2e37eab1bff50f727e979fcacb509e225fb853433a6fe21d2fb34e6305 29.54MB / 29.54MB              1.9s 
 => => sha256:2dde7cc6adac858452c1d6864cecd81f2294e0f91a055b66277de113772a699f 185B / 185B                    1.1s 
 => => sha256:ab866cbff8b4f2131ee423742e33a1804a7a7c3e5c93b9e4645af17b48975d69 1.61kB / 1.61kB                0.0s 
 => => sha256:a65b6fc69166e435211e2de3622b7bbf67cdca981977e7c0429af3adb1a66377 1.62kB / 1.62kB                0.0s 
 => => sha256:7fedf687405e2bd9ca6b5c52926fdc8bf32d7a05cba6a05b2035ae57996188d4 4.81kB / 4.81kB                0.0s 
 => => sha256:d32252c1bed431b062fdd8aadc224ac12b9a2a2eff90bf7052b9364711099bf1 152.21MB / 152.21MB           12.6s 
 => => sha256:067798a73ae61b70f499fd3d4819500cb3acd776e6faf232ab20bc872c91e8eb 95B / 95B                      1.7s 
 => => sha256:3779ca34757b8831444c435344684e390da90df7f511a7e36f613e938e24bfe2 10.69kB / 10.69kB              2.3s 
 => => sha256:912bb26dd00cc9c343e045930869bf98f8abf881a345e7cfc0e7ac58994d018a 6.25kB / 6.25kB                2.5s 
 => => extracting sha256:9cb31e2e37eab1bff50f727e979fcacb509e225fb853433a6fe21d2fb34e6305                     1.5s 
 => => sha256:caba482c969b09ff1c59771caaea8518da9213e369eb6ce042098bc02e9c2b25 151.66MB / 151.66MB           13.9s
 => => extracting sha256:d32252c1bed431b062fdd8aadc224ac12b9a2a2eff90bf7052b9364711099bf1                     5.9s
 => => extracting sha256:2dde7cc6adac858452c1d6864cecd81f2294e0f91a055b66277de113772a699f                     0.0s 
 => => extracting sha256:067798a73ae61b70f499fd3d4819500cb3acd776e6faf232ab20bc872c91e8eb                     0.0s 
 => => extracting sha256:3779ca34757b8831444c435344684e390da90df7f511a7e36f613e938e24bfe2                     0.0s 
 => => extracting sha256:912bb26dd00cc9c343e045930869bf98f8abf881a345e7cfc0e7ac58994d018a                     0.0s 
 => => extracting sha256:caba482c969b09ff1c59771caaea8518da9213e369eb6ce042098bc02e9c2b25                     6.0s 
 => [internal] load build context                                                                            29.3s 
 => => transferring context: 791.15MB                                                                        29.2s 
 => [rust 1/1] FROM docker.io/library/rust:1.85.0-bookworm@sha256:0ff31c9ffa641a62e48d543fb00b4960955ea375f  42.4s 
 => => resolve docker.io/library/rust:1.85.0-bookworm@sha256:0ff31c9ffa641a62e48d543fb00b4960955ea375f40776f  0.0s 
 => => sha256:0ff31c9ffa641a62e48d543fb00b4960955ea375f40776f40f585b89e654cc5e 7.75kB / 7.75kB                0.0s 
 => => sha256:16a7f242108de02f10fe4a392991679bafa7694e59f5b40a54d5af1be9b40d03 1.94kB / 1.94kB                0.0s 
 => => sha256:9436385238b1d9cddae43c7eb32bd9204021346f9ed6628c00c5ef55cbfc7cc0 4.36kB / 4.36kB                0.0s 
 => => sha256:7cd785773db44407e20a679ce5439222e505475eed5b99f1910eb2cda51729ab 48.47MB / 48.47MB             13.5s 
 => => sha256:091eb8249475f42de217265c501e0186f0a3ea7490ef7f51458c30db91fb3cac 24.01MB / 24.01MB             17.9s 
 => => extracting sha256:7cd785773db44407e20a679ce5439222e505475eed5b99f1910eb2cda51729ab                     2.6s 
 => => sha256:255774e0027b72d2327719e78dbad5ad8c9cf446d055e45be7fc149418470bae 64.40MB / 64.40MB             21.3s 
 => => sha256:353e14e5cc47664fba714a7da288001d90427c705494847ac773f5cc08199451 211.35MB / 211.35MB           28.1s 
 => => sha256:fe44f25b68dcc0fad03eddb07de055524bb7c31f088f3c1d411485751fba2b05 193.70MB / 193.70MB           39.9s 
 => => extracting sha256:091eb8249475f42de217265c501e0186f0a3ea7490ef7f51458c30db91fb3cac                     0.7s 
 => => extracting sha256:255774e0027b72d2327719e78dbad5ad8c9cf446d055e45be7fc149418470bae                     3.2s 
 => => extracting sha256:353e14e5cc47664fba714a7da288001d90427c705494847ac773f5cc08199451                     6.1s 
 => => extracting sha256:fe44f25b68dcc0fad03eddb07de055524bb7c31f088f3c1d411485751fba2b05                     2.4s 
 => [winarm64  2/10] RUN apt-get update &&     apt-get install -qqy --no-install-recommends g++ mingw-w64    85.1s 
 => [winx64 1/6] RUN apt-get update &&     apt-get install -qqy --no-install-recommends mingw-w64            32.5s 
 => [linuxarm64 1/6] RUN apt-get update &&     apt-get purge -y g++ &&     apt-get install -qqy --no-instal  17.2s 
 => [linuxx64 1/5] RUN rustup target add x86_64-unknown-linux-gnu                                             0.7s 
 => [stage-5  1/13] WORKDIR /src                                                                              0.3s 
 => [stage-5  2/13] RUN apt-get update &&     apt-get install -y nuget mono-complete dos2unix &&     wget h  62.8s 
 => [linuxx64 2/5] WORKDIR /src                                                                               0.0s 
 => [linuxx64 3/5] COPY . .                                                                                   0.8s 
 => [linuxx64 4/5] WORKDIR /src/rust                                                                          0.0s 
 => [linuxx64 5/5] RUN cargo build --target x86_64-unknown-linux-gnu --release                               22.2s 
 => [linuxarm64 2/6] RUN rustup target add aarch64-unknown-linux-gnu                                          7.8s 
 => [linuxarm64 3/6] WORKDIR /src                                                                             0.0s 
 => [linuxarm64 4/6] COPY . .                                                                                 1.0s 
 => [linuxarm64 5/6] WORKDIR /src/rust                                                                        0.0s 
 => [linuxarm64 6/6] RUN cargo build --target aarch64-unknown-linux-gnu --release                            18.4s 
 => [winx64 2/6] RUN rustup target add x86_64-pc-windows-gnu                                                  8.6s 
 => [winx64 3/6] WORKDIR /src                                                                                 0.0s 
 => [winx64 4/6] COPY . .                                                                                     1.1s 
 => [winx64 5/6] WORKDIR /src/rust                                                                            0.0s 
 => [winx64 6/6] RUN cargo build --target x86_64-pc-windows-gnu --release                                    20.9s 
 => [stage-5  3/13] COPY . .                                                                                  0.9s 
 => [stage-5  4/13] RUN dos2unix update_version.sh                                                            0.3s 
 => [stage-5  5/13] RUN bash update_version.sh                                                                0.7s 
 => [stage-5  6/13] COPY --from=winx64 /src/rust/target/x86_64-pc-windows-gnu/release/hf_tokenizers.dll /src  0.0s 
 => [winarm64  3/10] COPY --from=rust /usr/local/cargo /usr/local/cargo                                       0.0s 
 => [winarm64  4/10] COPY --from=rust /usr/local/rustup /usr/local/rustup                                     0.4s 
 => [winarm64  5/10] RUN rustup default stable                                                               34.6s 
 => [winarm64  6/10] RUN rustup target add aarch64-pc-windows-gnullvm                                         9.8s 
 => [winarm64  7/10] WORKDIR /src                                                                             0.0s 
 => [winarm64  8/10] COPY . .                                                                                 0.7s 
 => [winarm64  9/10] WORKDIR /src/rust                                                                        0.0s 
 => ERROR [winarm64 10/10] RUN cargo build --target aarch64-pc-windows-gnullvm --release                      8.9s 
------
 > [winarm64 10/10] RUN cargo build --target aarch64-pc-windows-gnullvm --release:
0.269 warning: unused manifest key: build
0.269 warning: unused manifest key: target.aarch64-unknown-linux-gnu.linker
0.295     Updating crates.io index
2.515  Downloading crates ...
3.002   Downloaded bitflags v1.3.2
3.033   Downloaded base64 v0.13.1
3.165   Downloaded ident_case v1.0.1
3.239   Downloaded number_prefix v0.4.0
3.262   Downloaded derive_builder_macro v0.20.0
3.310   Downloaded cfg-if v1.0.0
3.394   Downloaded darling_macro v0.20.9
3.803   Downloaded windows-targets v0.52.5
3.953   Downloaded fnv v1.0.7
4.143   Downloaded crossbeam-deque v0.8.5
4.474   Downloaded macro_rules_attribute-proc_macro v0.2.0
4.475   Downloaded macro_rules_attribute v0.2.0
4.477   Downloaded itoa v1.0.11
4.537   Downloaded monostate-impl v0.1.13
4.752   Downloaded monostate v0.1.13
4.753   Downloaded lazy_static v1.4.0
4.893   Downloaded strsim v0.11.1
4.903   Downloaded rand_core v0.6.4
4.911   Downloaded paste v1.0.15
4.917   Downloaded ppv-lite86 v0.2.17
4.930   Downloaded rayon-cond v0.3.0
4.937   Downloaded uuid-macro-internal v1.11.0
4.943   Downloaded thiserror-impl v1.0.61
4.953   Downloaded thiserror v1.0.61
4.966   Downloaded csbindgen v1.9.3
4.967   Downloaded console v0.15.8
4.970   Downloaded either v1.12.0
4.985   Downloaded darling v0.20.9
4.989   Downloaded getrandom v0.2.15
4.997   Downloaded derive_builder_core v0.20.0
5.022   Downloaded once_cell v1.19.0
5.024   Downloaded pkg-config v0.3.30
5.040   Downloaded rand_chacha v0.3.1
5.064   Downloaded quote v1.0.36
5.088   Downloaded derive_builder v0.20.0
5.100   Downloaded onig v6.4.0
5.106   Downloaded crossbeam-epoch v0.9.18
5.124   Downloaded crossbeam-utils v0.8.20
5.126   Downloaded encode_unicode v0.3.6
5.130   Downloaded log v0.4.21
5.148   Downloaded serde_derive v1.0.203
5.167   Downloaded unicode-ident v1.0.12
5.170   Downloaded smallvec v1.13.2
5.184   Downloaded uuid v1.11.0
5.186   Downloaded proc-macro2 v1.0.85
5.222   Downloaded darling_core v0.20.9
5.226   Downloaded indicatif v0.17.8
5.228   Downloaded ryu v1.0.18
5.282   Downloaded cc v1.0.99
5.329   Downloaded rayon-core v1.12.1
5.363   Downloaded unicode_categories v0.1.1
5.370   Downloaded serde v1.0.203
5.422   Downloaded rand v0.8.5
5.425   Downloaded unicode-segmentation v1.11.0
5.441   Downloaded unicode-normalization-alignments v0.1.12
5.465   Downloaded minimal-lexical v0.2.1
5.484   Downloaded memchr v2.7.4
5.540   Downloaded itertools v0.11.0
5.544   Downloaded nom v7.1.3
5.559   Downloaded itertools v0.12.1
5.587   Downloaded portable-atomic v1.6.0
5.616   Downloaded serde_json v1.0.117
5.653   Downloaded esaxx-rs v0.1.10
5.694   Downloaded aho-corasick v1.1.3
5.725   Downloaded rayon v1.10.0
5.735   Downloaded tokenizers v0.20.4
5.863   Downloaded regex v1.10.5
5.887   Downloaded syn v2.0.87
5.947   Downloaded regex-syntax v0.8.4
5.965   Downloaded unicode-width v0.1.13
5.971   Downloaded windows_aarch64_gnullvm v0.52.5
5.979   Downloaded spm_precompiled v0.1.4
5.985   Downloaded regex-automata v0.4.7
5.996   Downloaded onig_sys v69.8.1
6.010   Downloaded libc v0.2.155
6.032   Downloaded windows_x86_64_gnu v0.52.5
6.051   Downloaded windows-sys v0.52.0
6.140    Compiling proc-macro2 v1.0.85
6.140    Compiling unicode-ident v1.0.12
6.141    Compiling crossbeam-utils v0.8.20
6.142    Compiling cc v1.0.99
6.142    Compiling windows_aarch64_gnullvm v0.52.5
6.143    Compiling ident_case v1.0.1
6.143    Compiling strsim v0.11.1
6.145    Compiling fnv v1.0.7
6.148    Compiling memchr v2.7.4
6.148    Compiling cfg-if v1.0.0
6.151    Compiling libc v0.2.155
6.152    Compiling serde v1.0.203
6.152    Compiling regex-syntax v0.8.4
6.154    Compiling pkg-config v0.3.30
6.155    Compiling either v1.12.0
6.155    Compiling rayon-core v1.12.1
6.157    Compiling portable-atomic v1.6.0
6.157    Compiling paste v1.0.15
6.160    Compiling ppv-lite86 v0.2.17
6.161    Compiling encode_unicode v0.3.6
6.162    Compiling unicode-width v0.1.13
6.162    Compiling lazy_static v1.4.0
6.162    Compiling thiserror v1.0.61
6.178    Compiling getrandom v0.2.15
6.180    Compiling serde_json v1.0.117
6.189    Compiling minimal-lexical v0.2.1
6.262    Compiling ryu v1.0.18
6.292    Compiling windows-targets v0.52.5
6.297    Compiling itertools v0.11.0
6.298    Compiling unicode-segmentation v1.11.0
6.313    Compiling macro_rules_attribute-proc_macro v0.2.0
6.355    Compiling rand_core v0.6.4
6.367    Compiling windows-sys v0.52.0
6.370    Compiling smallvec v1.13.2
6.373    Compiling once_cell v1.19.0
6.390    Compiling itoa v1.0.11
6.417    Compiling number_prefix v0.4.0
6.440    Compiling bitflags v1.3.2
6.472    Compiling base64 v0.13.1
6.475    Compiling itertools v0.12.1
6.475    Compiling unicode_categories v0.1.1
6.480    Compiling rand_chacha v0.3.1
6.504    Compiling log v0.4.21
6.534    Compiling aho-corasick v1.1.3
6.548    Compiling nom v7.1.3
6.577    Compiling unicode-normalization-alignments v0.1.12
6.630    Compiling rand v0.8.5
6.708    Compiling crossbeam-epoch v0.9.18
6.723    Compiling quote v1.0.36
6.777    Compiling macro_rules_attribute v0.2.0
6.842    Compiling syn v2.0.87
6.875    Compiling onig_sys v69.8.1
6.876    Compiling esaxx-rs v0.1.10
6.888    Compiling crossbeam-deque v0.8.5
7.088    Compiling console v0.15.8
7.130 warning: esaxx-rs@0.1.10: clang: error: version 'llvm' in target triple 'aarch64-pc-windows-gnullvm' is invalid
7.130 error: failed to run custom build command for `esaxx-rs v0.1.10`
7.130
7.130 Caused by:
7.130   process didn't exit successfully: `/src/rust/target/release/build/esaxx-rs-c71f333768eea60d/build-script-build` (exit status: 1)
7.130   --- stdout
7.130   TARGET = Some("aarch64-pc-windows-gnullvm")
7.130   OPT_LEVEL = Some("3")
7.130   HOST = Some("x86_64-unknown-linux-gnu")
7.130   cargo:rerun-if-env-changed=CXX_aarch64-pc-windows-gnullvm
7.130   CXX_aarch64-pc-windows-gnullvm = None
7.130   cargo:rerun-if-env-changed=CXX_aarch64_pc_windows_gnullvm
7.130   CXX_aarch64_pc_windows_gnullvm = None
7.130   cargo:rerun-if-env-changed=TARGET_CXX
7.130   TARGET_CXX = None
7.130   cargo:rerun-if-env-changed=CXX
7.130   CXX = None
7.130   RUSTC_LINKER = None
7.130   cargo:rerun-if-env-changed=CROSS_COMPILE
7.130   CROSS_COMPILE = None
7.130   cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
7.130   cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
7.130   CRATE_CC_NO_DEFAULTS = None
7.130   DEBUG = Some("false")
7.130   cargo:rerun-if-env-changed=CXXFLAGS_aarch64-pc-windows-gnullvm
7.130   CXXFLAGS_aarch64-pc-windows-gnullvm = None
7.130   cargo:rerun-if-env-changed=CXXFLAGS_aarch64_pc_windows_gnullvm
7.130   CXXFLAGS_aarch64_pc_windows_gnullvm = None
7.130   cargo:rerun-if-env-changed=TARGET_CXXFLAGS
7.130   TARGET_CXXFLAGS = None
7.130   cargo:rerun-if-env-changed=CXXFLAGS
7.130   CXXFLAGS = Some("--stdlib=libc++")
7.130   cargo:warning=clang: error: version 'llvm' in target triple 'aarch64-pc-windows-gnullvm' is invalid        
7.130
7.130   --- stderr
7.130
7.130
7.130   error occurred: Command "aarch64-w64-mingw32-clang++" "-O3" "--target=aarch64-pc-windows-gnullvm" "-ffunction-sections" "-fdata-sections" "--target=aarch64-pc-windows-gnullvm" "-I" "src" "--stdlib=libc++" "-std=c++11" "-o" "/src/rust/target/aarch64-pc-windows-gnullvm/release/build/esaxx-rs-96aa1809e57f51bc/out/2e40c9e35e9506f4-esaxx.o" "-c" "src/esaxx.cpp" with args aarch64-w64-mingw32-clang++ did not execute successfully (status code exit status: 1).
7.130
7.130
7.130 warning: build failed, waiting for other jobs to finish...
7.176 warning: onig_sys@69.8.1: clang: error: version 'llvm' in target triple 'aarch64-pc-windows-gnullvm' is invalid
7.176 error: failed to run custom build command for `onig_sys v69.8.1`
7.176
7.176 Caused by:
7.176   process didn't exit successfully: `/src/rust/target/release/build/onig_sys-19cebd2a6fd95104/build-script-build` (exit status: 1)
7.176   --- stdout
7.176   cargo:rerun-if-env-changed=RUSTONIG_DYNAMIC_LIBONIG
7.176   cargo:rerun-if-env-changed=RUSTONIG_STATIC_LIBONIG
7.176   cargo:rerun-if-env-changed=RUSTONIG_SYSTEM_LIBONIG
7.176   TARGET = Some("aarch64-pc-windows-gnullvm")
7.176   OPT_LEVEL = Some("3")
7.176   HOST = Some("x86_64-unknown-linux-gnu")
7.176   cargo:rerun-if-env-changed=CC_aarch64-pc-windows-gnullvm
7.176   CC_aarch64-pc-windows-gnullvm = None
7.176   cargo:rerun-if-env-changed=CC_aarch64_pc_windows_gnullvm
7.176   CC_aarch64_pc_windows_gnullvm = None
7.176   cargo:rerun-if-env-changed=TARGET_CC
7.176   TARGET_CC = None
7.176   cargo:rerun-if-env-changed=CC
7.176   CC = None
7.176   RUSTC_LINKER = None
7.176   cargo:rerun-if-env-changed=CROSS_COMPILE
7.176   CROSS_COMPILE = None
7.176   cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
7.176   cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
7.176   CRATE_CC_NO_DEFAULTS = None
7.176   DEBUG = Some("false")
7.176   cargo:rerun-if-env-changed=CFLAGS_aarch64-pc-windows-gnullvm
7.176   CFLAGS_aarch64-pc-windows-gnullvm = None
7.176   cargo:rerun-if-env-changed=CFLAGS_aarch64_pc_windows_gnullvm
7.176   CFLAGS_aarch64_pc_windows_gnullvm = None
7.176   cargo:rerun-if-env-changed=TARGET_CFLAGS
7.176
7.176   --- stderr
7.176
7.176
7.176   error occurred: Command "aarch64-w64-mingw32-clang" "-O3" "--target=aarch64-pc-windows-gnullvm" "-ffunction-sections" "-fdata-sections" "--target=aarch64-pc-windows-gnullvm" "-I" "/src/rust/target/aarch64-pc-windows-gnullvm/release/build/onig_sys-ee3157a40f8448a9/out" "-I" "oniguruma/src" "-o" "/src/rust/target/aarch64-pc-windows-gnullvm/release/build/onig_sys-ee3157a40f8448a9/out/c77b18e714869709-regexec.o" "-c" "oniguruma/src/regexec.c" with args aarch64-w64-mingw32-clang did not execute successfully (status code exit status: 1).
7.176
7.176
------
Dockerfile:43
--------------------
  41 |
  42 |     ENV CXXFLAGS="--stdlib=libc++"
  43 | >>> RUN cargo build --target aarch64-pc-windows-gnullvm --release
  44 |
  45 |     # Linux X64
--------------------
ERROR: failed to solve: process "/bin/sh -c cargo build --target aarch64-pc-windows-gnullvm --release" did not complete successfully: exit code: 101
```

Some meaningful error log would be following?

```
7.130 warning: esaxx-rs@0.1.10: clang: error: version 'llvm' in target triple 'aarch64-pc-windows-gnullvm' is invalid
7.130 error: failed to run custom build command for `esaxx-rs v0.1.10`
7.130
7.130 Caused by:
7.130   process didn't exit successfully: `/src/rust/target/release/build/esaxx-rs-c71f333768eea60d/build-script-build` (exit status: 1)
```

```
7.130 warning: build failed, waiting for other jobs to finish...
7.176 warning: onig_sys@69.8.1: clang: error: version 'llvm' in target triple 'aarch64-pc-windows-gnullvm' is invalid
7.176 error: failed to run custom build command for `onig_sys v69.8.1`
```

sappho192 commented:
@Binozo 

By the way, I want to use PowerShell script as possible since I want to build this project using Windows machine in development. (Of course release build is better & should be built in Docker+Github Action)

So if possible, I'd like to suggest following rules and love to hear your opinion:
- When making cross-platform shell script, write in PowerShell
- If you are to make a shell script which should be run in specific OS only for compatibility issue, it is okay to use OS-native shell script

For example, shell script to update the version could be used both Docker and native environment when rapidly checking the development result, so I think it would be better to be written in cross-platform script language like PowerShell (or maybe Python?)
The script to pack the NuGet-stuff may _not_ be written in PowerShell since it would only be used in product-releasing step.

So in current Pull Request, I'd like to agree extracting updating version stuff as `update_version.ps1` like https://github.com/sappho192/Tokenizers.DotNet/pull/21/commits/a209699953c6b37dad9d56e7aac2cde446f1a7d7#diff-ffe8f742262246c4d60508e79ccf0cbb522c3ce1e004aea4409c34362f081d65 and using `pack.sh` for releasing the NuGet package.

Apart from above suggestions, I really appreciate all of your contributions.
It's always nice to have someone read my code, and new contributions and discussions around it makes me very motivating. 😊

Binozo commented:
Hi @sappho192 👋

To be honest I am really surprised about the error, that should not happen. Are you using  the newest Docker version?
Also I tested the Dockerfile on both Windows 11 x64 and Ubuntu Server x64 🤔
I hadn't the chance to test it on my primary Windows 11 arm64 workstation because of some updates Windows needs to release for WSL <-> Docker compatibility.

Could you please try the following commands on your machine? Maybe there is some git merging error from my side:

```sh
$ git clone https://github.com/Binozo/Tokenizers.DotNet
$ cd Tokenizers.DotNet
$ docker build -f Dockerfile -t ghcr.io/sappho192/tokenizers.dotnet:latest .
$ docker run -v .\out:/out --rm ghcr.io/sappho192/tokenizers.dotnet:latest
```

Additionally I support your decision to switch back to Powershell scripts for better cross platform compatibility, will send some new commits soon 😄

sappho192 commented:
> To be honest I am really surprised about the error, that should not happen. Are you using the newest Docker version?

Yes, I'm using latest Docker(v28.0.1, build 068a01e).

When I run the Dockerfile from your fork (https://github.com/Binozo/Tokenizers.DotNet), I could successfully build the project.

I've checked out this pull request from following UI, so maybe your fork and this PR has slight difference?
![image](https://github.com/user-attachments/assets/09e7a232-e1ee-4884-a4f2-4cccebc99b4f)

sappho192 commented:
Hi @Binozo with my apologies,
Just in case, I've once removed my local repo directory and clone again, and tried to checkout this PR & build the dockerfile.
The dockerfile has been built and run with no errors. 🤦‍♂️🫠

Very mysterious, but at least it was just my problem. So please go on forward ' v')b

sappho192 commented on update_version.sh:
As we discussed before, I'd like to suggest implementing this shell script in PowerShell.
Other changes seems alright for now 👍

Binozo replied:
Yes expect some new commits soon 🙏

sappho192 commented on README.md:
Does this `ghcr.io/sappho192/tokenizers.dotnet:latest` assume that we'll be putting the package on ghcr.io soon (which I think is fine)?

Binozo replied:
As it is a requirement from Docker to specify an Image Tag I thought `ghcr.io/sappho192/tokenizers.dotnet:latest` would fit best because it would _theoretically_ be the official image name if you would publish it on GitHub 🤔

Although I didn't expect it to be published I just wanted to make it compatible out-of-the box if it would be ever the case, additionally when people look though their saved Docker images (`docker image ls`) they instantly know where this image came from 🙂

sappho192 commented on pack.sh:
```diff
@@ -0,0 +1,36 @@
+
+# Retrieve the current Git branch and commit hash
+branch=$(git rev-parse --abbrev-ref HEAD)
+commit=$(git rev-parse HEAD)
```
Just out of curiosity, what does these two variables do? Those are not referenced in following code.
Are those variables reserved for future use?

Binozo replied:
Ah yes you're right, that are some leftovers

I had some problems with nuget in the Dockerfile, it didn't detect `$RepositoryBranch$` and `$RepositoryCommit$` right so at first I replaced them manually with `sed` but later I found out the installation script shipped a really old version of nuget, I think it was version 2.0.0 or so. Then I added `nuget update -self` and it worked without the need to replace anything. 🫠 I'll remove those obsolete variables

Binozo commented:
Glad to hear it worked 😄

Binozo commented:
Update 🚀🚀🚀

- Removed bash scripts
- Adjusted powershell scripts

Now by default the powershell script will prepare all `*.nuspec` files but only build for the current platform.

If you run the docker command from the README, docker will build for all platforms. No additional scripts or setup required ☺️

Binozo commented:
@sappho192 feel free to test this, I think this PR is now ready to merge 🙂

sappho192 commented:
I've done for some finalization stuff and now it looks good for merge. 
The final commit of this PR, 2aea642 is going to be tagged as 1.1.2 (but NuGet package will be released in 1.2.0 which new features will be added)

Thank you again @Binozo for your contribution 🤗

Binozo commented:
Hey @sappho192, fantastic! Thanks for the quick response times 😄

Have a nice week ☀️
